File: src/main/java/gr/ppdms/GoogleDriveDownloader.java
package gr.ppdms;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

public class GoogleDriveDownloader {

    public static String downloadFile(String fileURL, String destinationPath) throws IOException {
        // Clean up the file ID from the URL if necessary
        String fileId = extractFileId(fileURL);
        
        // Extract the resourcekey if available
        String resourceKey = extractResourceKey(fileURL);
        
        // Extract authuser if available
        String authUser = extractAuthUser(fileURL);
        
        String downloadUrl = String.format("https://drive.usercontent.google.com/download?id=%s&export=download", fileId);
        if (resourceKey != null) {
            downloadUrl += "&resourcekey=" + resourceKey;
        }
        if (authUser != null) {
            downloadUrl += "&authuser=" + authUser;
        }

        URL url = new URL(downloadUrl);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setUseCaches(false);
        conn.setRequestProperty("Accept-Encoding", "identity");
        conn.setRequestProperty("User-Agent", "Mozilla/5.0");
        
        destinationPath = destinationPath + "/" + getFileName(conn, fileURL);

        // Create the destination directory if it doesn't exist
        File destFile = new File(destinationPath);
        destFile.getParentFile().mkdirs();

        // Download with progress tracking
        try (InputStream in = conn.getInputStream();
             FileOutputStream out = new FileOutputStream(destinationPath)) {
            
            byte[] buffer = new byte[4096];
            int bytesRead;
            while ((bytesRead = in.read(buffer)) != -1) {
                out.write(buffer, 0, bytesRead);
            }
        }
        return destinationPath;
    }

    private static String extractFileId(String url) {
        // First try to match the file/d/ pattern
        String filePattern = "https://drive.google.com/file/d/([a-zA-Z0-9_-]+)";
        Pattern fileCompiled = Pattern.compile(filePattern);
        Matcher fileMatcher = fileCompiled.matcher(url);
        if (fileMatcher.find()) {
            return fileMatcher.group(1);
        }
        
        // Then try to match the open?id= pattern
        String openPattern = "id=([a-zA-Z0-9_-]+)";
        Pattern openCompiled = Pattern.compile(openPattern);
        Matcher openMatcher = openCompiled.matcher(url);
        if (openMatcher.find()) {
            return openMatcher.group(1);
        }
        
        return url;
    }

    private static String extractResourceKey(String url) {
        String pattern = "resourcekey=([^&]+)";
        Pattern compiledPattern = Pattern.compile(pattern);
        Matcher matcher = compiledPattern.matcher(url);
        if (matcher.find()) {
            return matcher.group(1);
        }
        return null;
    }

    private static String extractAuthUser(String url) {
        String pattern = "authuser=([^&]+)";
        Pattern compiledPattern = Pattern.compile(pattern);
        Matcher matcher = compiledPattern.matcher(url);
        if (matcher.find()) {
            try {
                return java.net.URLDecoder.decode(matcher.group(1), "UTF-8");
            } catch (IOException e) {
                return matcher.group(1);
            }
        }
        return null;
    }

    public static String getFileName(HttpURLConnection connection, String fileUrl) {
        String raw = connection.getHeaderField("Content-Disposition");
        
        // Check if the Content-Disposition header contains the filename
        if (raw != null && raw.contains("filename=")) {
            String fileName = raw.split("filename=")[1].trim();
            return fileName.replaceAll("\"", "");
        } else {
            // Fallback: Parse the HTML page to extract the title
            try {
                Document doc = Jsoup.connect(fileUrl).get();
                String title = doc.title();
                
                // Clean up the title if necessary (Google Drive adds "- Google Drive" to the title)
                if (title.contains("- Google Drive")) {
                    title = title.replace("- Google Drive", "").trim();
                }
                return title;
            } catch (IOException e) {
                e.printStackTrace();
                // Return a default name if an error occurs while fetching the page
                return "downloaded_file";
            }
        }
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.out.println("Usage: java GoogleDriveDownloader <fileUrl> <destinationPath>");
            return;
        }

        String fileUrl = args[0];
        String destinationPath = args[1];

        try {
            String savedPath = downloadFile(fileUrl, destinationPath);
            System.out.println("File downloaded successfully to: " + savedPath);
        } catch (IOException e) {
            System.err.println("Error downloading file: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: app/services/google_drive_downloader.py

import requests
import re
import sys
from typing import Optional
from urllib.parse import unquote
from bs4 import BeautifulSoup

def download_file(file_url, destination_path, webdav_uploader=None):
    """
    Downloads a file from a Google Drive URL and uploads to WebDAV.
    
    Args:
        file_url: The Google Drive URL
        destination_path: The destination WebDAV path
        webdav_uploader: WebDAV uploader instance (required)
        
    Returns:
        Tuple of (webdav_path, md5_hash)
    """
    from app.services.scraper import compute_md5, compute_md5_from_bytes
    
    file_id = extract_file_id(file_url)
    resource_key = extract_resource_key(file_url)
    auth_user = extract_auth_user(file_url)

    download_url = f"https://drive.usercontent.google.com/download?id={file_id}&export=download"
    if resource_key:
        download_url += f"&resourcekey={resource_key}"
    if auth_user:
        download_url += f"&authuser={auth_user}"

    with requests.get(download_url, stream=True) as r:
        r.raise_for_status()
        
        file_name = get_file_name(r, file_url)
        
        # WebDAV is required
        if not webdav_uploader or not webdav_uploader.is_configured():
            raise RuntimeError("WebDAV must be configured to download files")
        
        # Download to memory first, compute MD5, then upload
        file_data = b''
        for chunk in r.iter_content(chunk_size=8192):
            file_data += chunk
        
        md5_hash = compute_md5_from_bytes(file_data)
        webdav_path = webdav_uploader.upload_file(file_data, destination_path, file_name)
        
        return webdav_path, md5_hash

def extract_file_id(url):
    """
    Extracts the file ID from a Google Drive URL.
    """
    # Try to match the file/d/ pattern
    match = re.search(r"https://drive.google.com/file/d/([a-zA-Z0-9_-]+)", url)
    if match:
        return match.group(1)
    
    # Try to match the open?id= pattern
    match = re.search(r"id=([a-zA-Z0-9_-]+)", url)
    if match:
        return match.group(1)
        
    return url

def extract_resource_key(url):
    """
    Extracts the resource key from a Google Drive URL.
    """
    match = re.search(r"resourcekey=([^&]+)", url)
    if match:
        return match.group(1)
    return None

def extract_auth_user(url):
    """
    Extracts the authuser from a Google Drive URL.
    """
    match = re.search(r"authuser=([^&]+)", url)
    if match:
        return unquote(match.group(1))
    return None

def get_file_name(response, file_url):
    """
    Gets the file name from the Content-Disposition header or by parsing the HTML title.
    """
    content_disposition = response.headers.get('content-disposition')
    if content_disposition and 'filename=' in content_disposition:
        filename = content_disposition.split('filename=')[1].strip().strip('"')
        return filename
    else:
        try:
            page_content = requests.get(file_url).text
            soup = BeautifulSoup(page_content, 'html.parser')
            title = soup.title.string
            if "- Google Drive" in title:
                title = title.replace("- Google Drive", "").strip()
            return title
        except Exception as e:
            print(f"Error fetching page to get file name: {e}", file=sys.stderr)
            return "downloaded_file"

def main():
    if len(sys.argv) != 3:
        print("Usage: python google_drive_downloader.py <fileUrl> <destinationPath>")
        return

    file_url = sys.argv[1]
    destination_path = sys.argv[2]

    try:
        saved_path = download_file(file_url, destination_path)
        print(f"File downloaded successfully to: {saved_path}")
    except Exception as e:
        print(f"Error downloading file: {e}", file=sys.stderr)

if __name__ == "__main__":
    main()
